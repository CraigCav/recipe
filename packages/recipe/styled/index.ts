export {default as styled, useTheme, mq} from './styled';
