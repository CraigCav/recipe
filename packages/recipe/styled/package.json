{
  "name": "@recipe/styled",
  "main": "dist/main.js",
  "module": "dist/module.js",
  "types": "dist/types.d.ts",
  "source": "index.ts",
  "files": [
    "dist"
  ],
  "sideEffects": [
    "*.css"
  ],
  "peerDependencies": {
    "@emotion/core": "^10.0.27",
    "@emotion/styled": "^10.0.27",
    "react": "^16.8.0"
  }
}