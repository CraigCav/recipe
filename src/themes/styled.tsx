import styled, {CreateStyled, StyledOptions} from '@emotion/styled';
import {Interpolation} from '@emotion/core';
import * as standard from './standard';

const bps = {medium: '768px', large: '1061px'};

export function mq(bp: string, i: Interpolation): Interpolation {
  return {[`@media (min-width: ${bp in bps ? bps[bp] : bp})`]: i};
}

const wrapWithStandardTheme = interpolation => {
  if (typeof interpolation !== 'function') return interpolation;
  return ({theme, ...props}) =>
    interpolation({
      ...props,
      theme: Object.keys(theme).length ? theme : standard,
    });
};

const createStyled = (tag: any, options?: StyledOptions) => {
  const createStyledComponent = styled(tag, options);
  return function styledWithStandardTheme() {
    // eslint-disable-next-line prefer-rest-params
    const argsWithStandardTheme = [...arguments].map(wrapWithStandardTheme);
    const Component = createStyledComponent(...argsWithStandardTheme);
    // Tag component for analytics
    Object.assign(Component, {__recipe: process.env.npm_package_version});
    return Component;
  };
};

// avoid importing private @emotion/styled/src/tags by copying tags (so we can write styled.span etc)
const keys = Object.keys(createStyled);
Object.keys(styled)
  .filter(n => keys.indexOf(n) === -1)
  .forEach(tagName => {
    createStyled[tagName] = createStyled(tagName);
  });

export default createStyled as CreateStyled<typeof standard>;
