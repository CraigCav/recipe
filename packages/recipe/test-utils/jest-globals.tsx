import React from 'react';
import {render, cleanup} from '@testing-library/react';
import {ThemeProvider} from 'emotion-theming';
import {standard} from '../themes';

import '@testing-library/jest-dom/extend-expect';

const renderWithTheme = renderFn => (component, ...rest) =>
  renderFn(<ThemeProvider theme={standard}>{component}</ThemeProvider>, rest);

export const renderToHtml = renderWithTheme((ui, options) => {
  const {container} = render(ui, options);
  return container.outerHTML;
});

export const fullRender = renderWithTheme(render);

afterEach(cleanup);
