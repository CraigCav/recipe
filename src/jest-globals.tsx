import {shallow as enzymeShallow, render as enzymeRender, mount as enzymeMount} from 'enzyme';
import React from 'react';
import {create as reactCreate} from 'react-test-renderer';
import {render as rtlRender, cleanup} from 'react-testing-library';
import {renderToStaticMarkup} from 'react-dom/server';
import {ThemeProvider} from 'emotion-theming';
import {standard} from './themes';

import '@testing-library/jest-dom/extend-expect';

const renderWithTheme = renderFn => (component, ...rest) =>
  renderFn(<ThemeProvider theme={standard}>{component}</ThemeProvider>, rest);

export const shallow = tree => {
  const wrapper = enzymeShallow(<ThemeProvider theme={standard} />);
  const instance = wrapper.instance() as any;
  const context = instance.getChildContext();

  return enzymeShallow(tree, {context});
};

export const mount = tree => {
  const wrapper = enzymeShallow(<ThemeProvider theme={standard} />);
  const instance = wrapper.instance() as any;
  const context = instance.getChildContext();

  return enzymeMount(tree, {
    context,
    childContextTypes: ThemeProvider.childContextTypes,
  });
};

export const renderToHtml = renderWithTheme(renderToStaticMarkup);

export const render = renderWithTheme(enzymeRender);

export const fullRender = renderWithTheme(rtlRender);

export const create = renderWithTheme(reactCreate);

afterEach(cleanup);
