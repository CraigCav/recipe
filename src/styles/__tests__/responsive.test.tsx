import React from 'react';
import {responsive} from '..';
import styled from '../../themes/styled';
import {render} from '../../jest-globals';

describe('responsive()', () => {
  let fn;
  const theme = {
    breakpoints: {
      tablet: '37.5rem',
      desktop: '56.25rem',
      big: '75rem',
    },
  };
  const colors = {
    white: '#fff',
    black: '#000',
  };

  beforeEach(() => {
    fn = responsive('use', {
      light: 'color: #fff',
      dark: props => `color: ${props.colors.black}`,
    });
  });

  it('should return undefined if there is no matching prop', () => {
    expect(fn({})).toBeUndefined();
  });

  it('should render base styles plus media queries for each breakpoint provided', () => {
    const subjectUnderTest = fn({use: {base: 'light', desktop: 'dark'}, colors, theme});
    const Component = styled.div(subjectUnderTest);
    const actual = render(<Component />);
    expect(actual).toMatchSnapshot();
  });

  it('should raise an error if called without a base variant', () => {
    expect(() => fn({use: {desktop: 'dark'}, colors, theme})).toThrow();
  });

  it('should raise an error if called without a theme', () => {
    expect(() => fn({use: {base: 'light', desktop: 'dark'}, colors})).toThrow();
  });

  it('should raise an error if called without breakpoints', () => {
    expect(() => fn({use: {base: 'light', desktop: 'dark'}, colors, theme: {}})).toThrow();
  });
});
