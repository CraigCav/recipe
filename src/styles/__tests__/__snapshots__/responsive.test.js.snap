// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`responsive() should render base styles plus media queries for each breakpoint provided 1`] = `
.recipe-0 {
  color: #fff;
}

@media screen and (min-width:56.25rem) {
  .recipe-0 {
    color: #000;
  }
}

<div
  class="recipe-0 recipe-1"
/>
`;
